### PR TITLE
GEOMESA-2172 Improved support for wide schemas

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Support for column groups in HBase and Accumulo through attribute user data
* Minimal column group is selected as query time to reduce overall data being read
* Query transforms are applied to BIN, density and stats scans based on the fields needed
* JsonPathPropertyAccessor updated to return AttributeDescriptors
  * Allows for determining properties read in JsonPath expressions
* QueryPlanFilterVisitor updated to handle type binding
  * Required to not bind JsonPath expressions to String
* Stats updated to store attribute name instead of index
  * Allows for use of column groups in stats scans, as exact schema being read doesn't matter
* SimpleFeatureType parsing updated to allow for quoted attribute user data values
* HBase binding changed from Query to Scan to support `addColumn`
* HBaseQueryPlan range configuration moved to HBasePlatform to allow for column families
* Updated examples to use SchemaBuilder instead of deprecated SftBuilder

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 1]
```diff
@@ -71,73 +68,54 @@ object HBaseQueryPlan {
...
-    val (scan, filterList) = calculateScanAndFilterList(ranges, remoteFilters)
     val hbaseTable = ds.connection.getTable(table)
-
-    ds.applySecurity(scan)
-
-    val results = GeoMesaCoprocessor.execute(hbaseTable, scan, filterList, coprocessorConfig.options).collect {
+    val results = GeoMesaCoprocessor.execute(hbaseTable, coprocessorScan, coprocessorConfig.options).collect {
       case r if r.size() > 0 => coprocessorConfig.bytesToFeatures(r.toByteArray)
     }
+//    val byteArray = serializeOptions(coprocessorConfig.configureScanAndFilter(coprocessorScan))
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Commented code?  Can it go?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBasePlatform.scala [hunk 2]
```diff
@@ -23,49 +23,51 @@ trait HBasePlatform extends HBaseIndexAdapter {
...
                                                toFeatures: (Iterator[Result]) => Iterator[SimpleFeature]): HBaseQueryPlan = {
+    val filterList = hbaseFilters.sortBy(_._1).map(_._2)
     coprocessor match {
       case None =>
         // optimize the scans
-        val scans = ranges.head match {
-          case t: Get  => configureGet(ranges.asInstanceOf[Seq[Get]], hbaseFilters)
-          case t: Scan => configureMultiRowRangeFilter(ds, ranges.asInstanceOf[Seq[Scan]], hbaseFilters)
+        val scans = if (ranges.head.isSmall) {
+          configureGet(ds, ranges, colFamily, filterList)
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBasePlatform.scala]: Rename method?

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+package org.locationtech.geomesa.index.planning
+
+import org.geotools.process.vector.TransformProcess
+import org.geotools.process.vector.TransformProcess.Definition
+import org.locationtech.geomesa.filter.FilterHelper
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.filter.Filter
+
+object Transforms {
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala]: Comment or rename the class:)?  What kind of transforms?!?!??!?!

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    names.forall(sft.indexOf(_) != -1)
+  }
+
+  /**
+    * Parses transform string into definitions
+    *
+    * @param transforms transform definitions
+    * @return
+    */
+  def definitions(transforms: String): Seq[Definition] = TransformProcess.toDefinition(transforms).asScala
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala]: From a discussion with GeoTools/GeoServer devs, it was suggested to look at the the GeoTools TransformFeatureStore or FeatureCollection.

**Comment by jnh5y** [comment]
@jahhulbert-ccri I think this is good to go after a few quick updates.  Do you want to review?  (Silence means no...:))

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: good catch, thanks

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBasePlatform.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBasePlatform.scala]: renamed

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala]: added comment

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala (hunk 4 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/Transforms.scala]: opened a ticket here: https://geomesa.atlassian.net/browse/GEOMESA-2321

**Pull request merged**